src/Form/AuthenticationForm.php

```diff
@@ -136,622 +105,88 @@ public function getFormId() {
-      // It's and edge case, set the existing key entity as an active key.
-      $this
-        ->config(static::CONFIG_NAME)
-        ->set('active_key', $new_key_id)
-        ->save();
-    }
-
-    // Create a new key.
-    $new_key = Key::create([
-      'id' => $new_key_id,
-      'label' => $this->t('Apigee Edge connection'),
-      'description' => $this->t('Contains the credentials for connecting to Apigee Edge.'),
-      'key_type' => 'apigee_auth',
-      'key_input' => 'apigee_auth_input',
-      'key_provider' => 'apigee_edge_private_file',
-    ]);
-    $new_key->save();
-
-    try {
-      /** @var \Drupal\apigee_edge\Plugin\KeyProviderRequirementsInterface $key_provider */
-      if (($key_provider = $new_key->getKeyProvider()) instanceof KeyProviderRequirementsInterface) {
-        $key_provider->checkRequirements($new_key);
-        // Write out an empty key.
-        $key_provider->setKeyValue($new_key, json_encode((object) ['auth_type' => EdgeKeyTypeInterface::EDGE_AUTH_TYPE_BASIC]));
+      catch (KeyProviderRequirementsException $exception) {
+        $key_provider_requirements_error = TRUE;
+        $form['settings']['provider_section']['key_provider']['#attributes']['class'][] = 'error';
       }
     }
-    catch (KeyProviderRequirementsException $exception) {
-      // Do nothing here, displaying error messages is handled in buildForm().
-    }
-
-    // Save the active key.
-    $this
-      ->config(static::CONFIG_NAME)
-      ->set('active_key', $new_key_id)
-      ->save();
-
-    return $new_key;
-  }
-
-  /**
-   * Moves form errors from one form state to another.
-   *
-   * Copied from `\Drupal\key\Form\KeyFormBase::moveFormStateErrors()`.
-   *
-   * @param \Drupal\Core\Form\FormStateInterface $from
-   *   The form state object to move from.
-   * @param \Drupal\Core\Form\FormStateInterface $to
-   *   The form state object to move to.
-   */
-  protected function moveFormStateErrors(FormStateInterface $from, FormStateInterface $to) {
-    foreach ($from->getErrors() as $name => $error) {
-      $to->setErrorByName($name, $error);
-    }
-  }
-
-  /**
-   * Moves storage variables from one form state to another.
-   *
-   * Copied from `\Drupal\key\Form\KeyFormBase::moveFormStateStorage()`.
-   *
-   * @param \Drupal\Core\Form\FormStateInterface $from
-   *   The form state object to move from.
-   * @param \Drupal\Core\Form\FormStateInterface $to
-   *   The form state object to move to.
-   */
-  protected function moveFormStateStorage(FormStateInterface $from, FormStateInterface $to) {
-    foreach ($from->getStorage() as $index => $value) {
-      $to->set($index, $value);
-    }
-  }
+    $form['settings']['provider_section']['#open'] = $key_provider_requirements_error || ($form_state->getTriggeringElement() && $form_state->getTriggeringElement()['#name'] === 'key_provider');
+    $form['settings']['provider_section']['#weight'] = $form['settings']['input_section']['#weight'] + 1;
```

Comment:
Is there a reason we are moving this under the `input_section`?

Reply:
Apigee did not want to expose the Key provider selector on this form originally to keep it simple and speed up the first-time configuration for those who just would like to try to out the module. Although this implementation hid the fact that the Authentication form saves the Apigee Edge credentials to an unencrypted file in the private file system by default. Different Key provider could have been configured only on the Key module's form, which could have been found only by those ppl who know the module or read its documentation. I always felt this is a bad idea because ppl do not read documentations and they will use the private key provider in production sites as well.

Now the current UI combines the best from both. The credential configuration form still has the primary focus on the top of the page but the key provider selector is also visible, only in the bottom of the page. Any user can easily find it, they do not need to navigate to a different UI to select more secure credential storage.

Reply:
I like the idea of getting the “best from both”!  Good solution to have advanced settings, this makes it more clear to the user that there are other options and to understand the tie to the Key module in a more clear way.